main/src/ca/uwaterloo/flix/language/phase/typer/TypeReduction.scala

```diff
@@ -274,59 +275,80 @@ object TypeReduction {
-    val candidates = candidateMethods(clazz, methodName, ts, static)
-
-    candidates match {
-      case Nil => JavaMethodResolution.NotFound
-      case method :: Nil => JavaMethodResolution.Resolved(method)
-      case _ :: _ :: _ =>
-        // Multiple candidate constructors exist according to subtyping, so we search for an exact
-        // match. Candidates could contain `append(String)` and `append(Object)` for the call
-        // `append("a")`.
-        val exactMatches = candidates.filter(m => exactArguments(m.getParameterTypes, ts))
-
-        exactMatches match {
-          // No exact matches. We have ambiguity among the candidates.
-          case Nil => JavaMethodResolution.AmbiguousMethod(candidates)
-
-          case exact :: Nil => JavaMethodResolution.Resolved(exact)
-
-          // Multiple exact matches are impossible in Java.
-          case _ :: _ :: _ =>
-            throw InternalCompilerException("Unexpected multiple exact matches for Java method", loc)
-        }
-    }
-  }
-
-  /**
-    * Removes methods of super-classes that are overridden with the same types.
-    */
-  private def candidateMethods(clazz: Class[?], methodName: String, ts: List[Type], static: Boolean): List[Method] = {
-    // this list contains e.g. both `StringBuilder.appendCodePoint(int)` and `AbstractStringBuilder.appendCodePoint(int)`.
-    val allCandidates = JvmOps.getMethods(clazz).filter(isCandidateMethod(_, methodName, static, ts))
-
-    def notOverridden(method: Method): Boolean = {
-      // this is a very hardcoded hack to make the standard library compile
-      // conceptually we should whether the defining classes are subtypes of eachother with exact signatures
-      // but that requires the Class object, which getMethods don't give us
-      method.getReturnType.equals(Void.TYPE) ||
-        method.getReturnType.isPrimitive ||
-        method.getReturnType.isArray ||
-        java.lang.reflect.Modifier.isInterface(method.getReturnType.getModifiers) ||
-        !(java.lang.reflect.Modifier.isAbstract(method.getReturnType.getModifiers) && !static)
+    val tparams = ts.map(getJavaType)
+    val m = MethodUtils.getMatchingAccessibleMethod(clazz, methodName, tparams *)
+    // We check if we found a method and if its static flag matches.
+    if (m != null && JvmOps.isStatic(m) == static) {
+      // Case 1: We found the method on the clazz.
+      JavaMethodResolution.Resolved(m)
+    } else {
+      // Case 2: We failed to find the method on the clazz.
+      // We make one attempt on java.lang.Object.
+      val classObj = classOf[java.lang.Object]
+      val m = MethodUtils.getMatchingAccessibleMethod(classObj, methodName, tparams *)
+      if (m != null) {
+        // Case 2.1: We found the method on java.lang.Object.
+        JavaMethodResolution.Resolved(m)
+      } else {
+        // Case 2.2: We failed to find the method, so we report an error on the original clazz.
+        JavaMethodResolution.NotFound
```

Comment:
We can no longer return ambiguous method errors because either we find the method or we don't.

Reply:
Thus both JavaMethodResolution and TypeError can be cleaned up.